werkit/compute/graph/_dependency_graph.py

```diff
@@ -0,0 +1,168 @@
+        else:
+            # TODO: Perhaps catch and re-throw to improve the error message.
+            return t.cast(t.Type[CustomType], self.value_type).coerce(value)
+
+    def serialize(self, value: t.Any) -> JSONType:
+        if self.value_type_is_built_in:
+            return value
+        else:
+            value_type = t.cast(t.Type[CustomType], self.value_type)
+            serialized = value_type.serialize(value)
+            value_type.validate(serialized)
+            return serialized
+
+
+InputJSONType = TypedDict("InputJSONType", {"valueType": str})
+
+
+class Input(BaseNode):
+    def as_native(self) -> InputJSONType:
+        return {"valueType": self.value_type_name}
+
+
+ComputeNodeJSONType = TypedDict(
+    "ComputeNodeJSONType", {"valueType": str, "dependencies": t.List[str]}
+)
+
+
+class ComputeNode(BaseNode):
+    # TODO: Use a narrower type for `method`.
+    def __init__(self, method: t.Callable, value_type: AnyValueType):
+        super().__init__(value_type=value_type)
+
+        self.method = method
+        self.dependencies = [x for x in inspect.signature(method).parameters.keys()][1:]
+
+    def bind(self, instance: ComputeNode) -> t.Callable:
+        import functools
+
+        return functools.partial(self.method, instance)
+
+    def as_native(self) -> ComputeNodeJSONType:
+        return {
+            "valueType": self.value_type_name,
+            "dependencies": self.dependencies,
+        }
+
+
+class Intermediate(ComputeNode):
+    pass
+
+
+class Output(ComputeNode):
+    pass
+
+
+def intermediate(value_type: AnyValueType):
+    def decorator(method):
+        return Intermediate(method, value_type)
+
+    return decorator
+
+
+def output(value_type: AnyValueType):
+    def decorator(method):
+        return Output(method, value_type)
+
+    return decorator
+
+
+DependencyGraphJSONType = TypedDict(
+    "DependencyGraphJSONType",
+    {
+        "schemaVersion": Literal[1],
+        "inputs": t.Dict[str, InputJSONType],
+        "intermediates": t.Dict[str, ComputeNodeJSONType],
+        "outputs": t.Dict[str, ComputeNodeJSONType],
+    },
+)
+
+AttrType = t.TypeVar("AttrType")
+
+
+def _attrs_of_type(obj: t.Any, _type: t.Type[AttrType]) -> t.Dict[str, AttrType]:
+    return {
+        name: getattr(obj, name)
+        for name in dir(obj)
+        if not name.startswith("__") and isinstance(getattr(obj, name), _type)
+    }
+
+
+class DependencyGraph:
+    def __init__(self, cls: t.Type):
+        assert inspect.isclass(cls)
+
+        self.inputs = _attrs_of_type(cls, Input)
+        self.intermediates = _attrs_of_type(cls, Intermediate)
+        self.outputs = _attrs_of_type(cls, Output)
+        self.compute_nodes = dict(**self.intermediates, **self.outputs)
+        self.all_nodes = dict(**self.inputs, **self.compute_nodes)
+
+    def keys(self) -> t.List[str]:
+        return list(self.inputs.keys()) + list(self.compute_nodes.keys())
+
+    def as_native(self) -> DependencyGraphJSONType:
```

Comment:
I found the method name `as_native` confusing. To me, this made me think of serializing to a "canonical" or "built-in" type - something "native" to python. Instead, this seems to want to generate the JSON data structure which is a symbolic representation of the dependency graph, for consumption in other applications that will consume the output of the python compute graph. So I wonder if we could call this `to_model`, `to_graph`, `to_graph_json`, or `to_serializable_graph`.

Reply:
Ah, yes, makes sense. I think `serialize` might make sense; that seems to be what I'm calling this in all the rest of the compute graph.